[CALCITE-2914] Improve how LatticeSuggester deduces foreign keys

julianhyde opened the pull request:
See https://issues.apache.org/jira/browse/CALCITE-2914

hsyuan commented on core/src/main/codegen/templates/Parser.jj:
```diff
@@ -5652,6 +5652,24 @@ SqlCall withinGroup(SqlNode arg) :
+        }
+    )
+}
+
```
Isn't this part (RESPECT/IGNORE NULLS) already in master by commit: https://github.com/apache/calcite/commit/d7946a94adfd2e788f5d324910944dd65dab11ee ?

chunweilei replied:
There are 3 commits in this PR. It looks like the first commit  [0c0e90](https://github.com/apache/calcite/pull/1141/commits/0c0e908b64949095af1c9c21815880089f52c34f) is the very commit.

danny0405 replied:
Yes, multiple PRS are here

julianhyde replied:
I pushed the first 2 commits to master very soon after opening this PR. The delta between master and this PR is therefore just one commit, 0c0e908b64949095af1c9c21815880089f52c34f. I assumed that github would figure that out, but it didn't. Please just consider that commit.

chunweilei commented on core/src/main/java/org/apache/calcite/materialize/SqlStatisticProvider.java:
```diff
@@ -16,16 +16,41 @@
+   * <p>For example, {@code isKey(EMP, [DEPTNO]} returns true;
+   * <p>For example, {@code isKey(DEPT, [DEPTNO]} returns false. */
+  boolean isKey(RelOptTable table, List<Integer> columns);
 }
```
Maybe we can use `isPrimaryKey` since there is `isForeignKey`.

julianhyde replied:
A table can have several unique keys, at most one of which is made the primary key. This algorithm can detect unique keys (also known as candidate keys) but has no way of knowing whether a key is the primary key. Rather than 'unique key' I went with just 'key'.

chunweilei replied:
Got it.

chunweilei commented on core/src/main/java/org/apache/calcite/statistic/QuerySqlStatisticProvider.java:
```diff
@@ -0,0 +1,192 @@
+               ResultSet resultSet = statement.executeQuery(sql)) {
+            if (!resultSet.next()) {
+              throw new AssertionError();
+            }
```
Maybe we can leave some error message when throwing the error.

julianhyde replied:
I agree. Done.

danny0405 commented on core/src/main/java/org/apache/calcite/materialize/SqlStatisticProvider.java:
```diff
@@ -16,16 +16,41 @@
 
 /**
  * Estimates row counts for tables and columns.
  *
```
The comments are not that accurate, we should add isKey and isForeignKey

julianhyde replied:
I agree. Done.

danny0405 commented on core/src/main/java/org/apache/calcite/statistic/QuerySqlStatisticProvider.java:
```diff
@@ -0,0 +1,192 @@
+          //   SELECT COUNT(*) FROM (
+          //     SELECT deptno FROM `EMP` WHERE deptno IS NOT NULL
+          //     MINUS
+          //     SELECT deptno FROM `DEPT`)
```
We also need to check that the toColumn are primary key of toTable

julianhyde replied:
By design, the `isForeignKey` method only checks referential integrity - i.e. that there is at least one matching row. It is left to `isKey` to determine "at most one".

danny0405 commented on core/src/main/java/org/apache/calcite/statistic/MapSqlStatisticProvider.java:
```diff
@@ -0,0 +1,138 @@
+ * table.
+ *
+ * <p>Only for testing.
+ */
```
We should move it to test source root if it is only used for testing.

julianhyde replied:
Not really possible. It is used outside of the core module.

danny0405 commented on core/src/main/java/org/apache/calcite/statistic/MapSqlStatisticProvider.java:
```diff
@@ -0,0 +1,138 @@
+        && keyMap.get(table.getQualifiedName().toString())
+            .equals(
+                ImmutableList.of(table.getRowType().getFieldList()
+                  .get(columns.get(0)).getName()));
```
We should pre-check for indexOutofBounds for ImmutableList.of(table.getRowType().getFieldList().get(columns.get(0))

julianhyde replied:
I changed it to fail the test if any of the columns are out of range. Also it now allows composite keys (even though none of those currently defined are composite).

danny0405 commented on core/src/main/java/org/apache/calcite/tools/Frameworks.java:
```diff
@@ -38,6 +37,7 @@
 import org.apache.calcite.sql2rel.SqlToRelConverter;
 import org.apache.calcite.sql2rel.StandardConvertletTable;
+import org.apache.calcite.statistic.MapSqlStatisticProvider;
 import org.apache.calcite.util.Util;
```
Why we use a test MapSqlStatisticProvider for Frameworks

julianhyde replied:
That was a mistake. It was useful in tests, so we didn't notice that we'd made it the default. I have now made a query provider wrapped in a caching provider the default.

julianhyde commented:
Thanks for the reviews. I have addressed all comments in efbefb1. Can someone re-review? I plan to push in about 12 hours if there are no further objections.
